### PR TITLE
Branch for Suzanne review 26 June 2018

### PR DESCRIPTION
**Pull request by julesnuggy**
Programme currently:
* Grabs locations.json file and extracts the list of working sensors to SensorList.
* Makes SQS subscription to get messages and pulls a message every second.
* If the message contains a sensor which is working, it will display the message.
* If the the sensor is not working, it will say so.

Programme issues:
* Programme is currently set to loop for 3600 secs (1 hr) - need to sort out mechanism to enable manual forced quit such that the queue is still destroyed. Would recommend setting the `if (counter > X)` statement in Main line 54 to a more manageable number before running.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
 
 target/
 
 src/main/resources/liquibase.properties
 
-dependency-reduced-pom.xml
\ No newline at end of file
+dependency-reduced-pom.xml
+
+locations.json
```
Comment (suzannehamilton) [inline .gitignore]: It looks like locations.json is still checked in, even though it's (correctly) in .gitignore.

The same goes for the log file.

File: src/main/resources/log4j2.xml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Configuration status="warn" name="SupportBank" packages="">
+    <Appenders>
+        <File name="FileAppender" fileName="logs/eventprocessing.log">
```
Comment (suzannehamilton) [inline src/main/resources/log4j2.xml]: You might find it helpful to have a `ConsoleAppender` as well as a `FileAppender`, so you can see the logs in the IntelliJ console.

File: src/main/java/eventprocessing/storage/MessageLog.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+public class MessageLog {
+
+    private List<String> messageHistory = new ArrayList<>();
+    private int maxSize;
+
+    public MessageLog(int maxSize) {
+        this.maxSize = maxSize;
+    }
+
+    public List<String> getMessageHistory() {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/storage/MessageLog.java]: Be careful when returning a collection of objects for other parts of your code to modify - it's really easy to lose track of what modifies the list and when.

Having a `truncate` method in this class is a good idea - how could you take the idea further? Is it possible to get rid of the `getMessageHistory` method entirely, so this class is the only thing that modifies the list?

File: src/main/java/eventprocessing/storage/MessageLog.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package eventprocessing.storage;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class MessageLog {
+
+    private List<String> messageHistory = new ArrayList<>();
+    private int maxSize;
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/storage/MessageLog.java]: If you make both of these fields final, it makes it clear they're never reassigned.

File: src/main/java/eventprocessing/responseservices/ResponseService.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package eventprocessing.responseservices;
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+import eventprocessing.models.Message;
+import eventprocessing.models.Response;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+public class ResponseService {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/responseservices/ResponseService.java]: `Response` and `ResponseService` sound very general, but it looks like they handle a particular kind of AWS API response. Have a think about the names to make it clearer.

File: src/main/java/eventprocessing/responseservices/ResponseService.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        try {
+            responseObject = gson.fromJson(jsonString, Response.class);
+            String messageString = responseObject.getMessageString();
+            Message messageObject = gson.fromJson(messageString, Message.class);
+            responseObject.setMessage(messageObject);
+        } catch (IllegalStateException | JsonSyntaxException e) {
+            logger.warn("Invalid JSON string received" + e.getMessage());
+            logger.warn("Received json string: " + jsonString);
+        }
+        return responseObject;
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/responseservices/ResponseService.java]: This returns null if there were errors, which looks like it will cause a `NullPointerException` elsewhere in your code, and you'll lose the information about the actual failure.

How else could you handle the error? Give me a shout if you want to talk about the options.

File: src/main/java/eventprocessing/responseservices/ResponseService.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    public Response parseResponse(String jsonString) {
+        Gson gson = new Gson();
+        Response responseObject = null;
+        try {
+            responseObject = gson.fromJson(jsonString, Response.class);
+            String messageString = responseObject.getMessageString();
+            Message messageObject = gson.fromJson(messageString, Message.class);
+            responseObject.setMessage(messageObject);
+        } catch (IllegalStateException | JsonSyntaxException e) {
+            logger.warn("Invalid JSON string received" + e.getMessage());
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/responseservices/ResponseService.java]: Logging tip: all the log4j logging methods (`error`, `warn`, etc.) have an overload which takes a `Throwable` as well as a message, for example: https://logging.apache.org/log4j/2.0/log4j-api/apidocs/org/apache/logging/log4j/Logger.html#error(org.apache.logging.log4j.message.Message,%20java.lang.Throwable)

This gives you better logs, because it includes the stack trace as well as the message.

File: src/main/java/eventprocessing/responseservices/ResponseService.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    private final Logger logger = LogManager.getLogger("ResponseService");
+
+    public Response parseResponse(String jsonString) {
+        Gson gson = new Gson();
+        Response responseObject = null;
+        try {
+            responseObject = gson.fromJson(jsonString, Response.class);
+            String messageString = responseObject.getMessageString();
+            Message messageObject = gson.fromJson(messageString, Message.class);
+            responseObject.setMessage(messageObject);
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/responseservices/ResponseService.java]: This is quite hard to follow - is it possible to deserialize everything at once, so you don't have to call this setter directly?

**Review by suzannehamilton — COMMENTED** [review]
I'm still looking through the code, but I've posted a few thoughts on things I've spotted so far. 🙂

File: src/main/java/eventprocessing/amazonservices/SnsClient.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+public class SnsClient {
+    private AmazonSNS sns;
+    private final static Logger logger = LogManager.getLogger("SnSClient");
+
+    public AmazonSNS buildSNSClient() {
+        try {
+            this.sns = AmazonSNSClient
+                    .builder()
+                    .withRegion(System.getenv("AWS_REGION"))
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/amazonservices/SnsClient.java]: This should probably be in `S3Details` for consistency - there's nothing secret about it, and it needs to be the same as the region in the topic ARN.

File: src/main/java/eventprocessing/amazonservices/SnsClient.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+                    .withCredentials(new EnvironmentVariableCredentialsProvider())
+                    .build();
+            logger.info("SnsClient created");
+        } catch (SdkClientException e) {
+            logger.error("Unable to create AmazonSNSClient");
+            logger.error(e.getMessage());
+        }
+        return sns;
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/amazonservices/SnsClient.java]: See the comment further down about returning null when there's an error. What do you _want_ your program to do if it can't build an SNS client, and is returning null the best way to do that?

File: src/main/java/eventprocessing/amazonservices/AmazonController.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import com.amazonaws.services.sns.util.Topics;
+import eventprocessing.DataReader;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+
+public class AmazonController {
+
+    private static final Logger logger = LogManager.getLogger("AmazonController");
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/amazonservices/AmazonController.java]: It's not a big deal, but it's better to pass `AmazonController.class` rather than a string when creating the logger, as you've done elsewhere. Then if you rename the class, the log name will get updated too.

File: src/main/java/eventprocessing/amazonservices/S3Details.java [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package eventprocessing.amazonservices;
+
+public class S3Details {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/amazonservices/S3Details.java]: Nitpick: the class is called `S3Details` but it contains some non-S3 config.

File: src/main/java/eventprocessing/amazonservices/S3Client.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.S3ObjectInputStream;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class S3Client {
+    Logger logger = LogManager.getLogger("S3Client");
+
+    public AmazonS3 clientBuilder() {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/amazonservices/S3Client.java]: Could this method be private?

File: src/main/java/eventprocessing/amazonservices/S3Client.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+    public S3ObjectInputStream generateS3InputStream() {
+        AmazonS3 s3 = clientBuilder();
+        S3ObjectInputStream s3is = null;
+        try {
+            S3Object o = s3.getObject(S3Details.s3BucketLocation, S3Details.s3Key);
+            s3is = o.getObjectContent();
+        } catch (AmazonServiceException e) {
+            System.err.println(e.getErrorMessage());
+            System.exit(1);
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/amazonservices/S3Client.java]: It's a bit scary for a class like `S3Client` to have a `System.exit` in it. I'd expect a decision to stop the application to be in some higher level control logic, such as in `main` itself.

File: src/main/java/eventprocessing/Main.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    private static Scanner scanner;
+    private static AmazonController amazonController;
+    private static ResponseList responseList;
+
+    public static void main(String[] args) throws IOException, InterruptedException {
+        logger.debug("App launched");
+        createObjects();
+        JSONParser jsonParser = new JSONParser("locations.json");
+        SensorList sensorList = jsonParser.createSensorList();
+        sensorList.getSensors().forEach(sensor -> System.out.println(sensor.getId()));
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/Main.java]: There's a mix of `logger` and `System.out` here. log4j gives you a lot of control over where the logs appear, so you should log everything to it. You could configure it to, for example, log everything `DEBUG` and higher to a file but just show `WARN` and above in the console.

Same goes for the other `System.out` calls elsewhere.

File: src/main/java/eventprocessing/Main.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+                    System.out.println(msg.getBody());
+                    responseList.getResponses().add(response);
+                    messageLog.getMessageHistory().add(response.messageId);
+                    messageLog.truncateIfExceedsMaxSize();
+                } else {
+                    System.out.println("sensor not working with id: " + response.getMessage().getLocationId());
+                }
+            }
+            System.out.println("=================================================");
+            Thread.sleep(1000);
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/Main.java]: Have a look into the blocking request that we talked about earlier to see if you can get rid of this `sleep`.

File: src/main/java/eventprocessing/DataReader.java [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+public class DataReader {
+    public void saveToFile(S3ObjectInputStream amazonS3Stream) throws IOException {
+        FileOutputStream fos = new FileOutputStream(new File("locations.json"));
+        byte[] read_buf = new byte[1024];
+        int read_len;
+        while ((read_len = amazonS3Stream.read(read_buf)) > 0) {
+            fos.write(read_buf, 0, read_len);
+        }
+        amazonS3Stream.close();
+        fos.close();
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/DataReader.java]: These `close` lines won't be called if an exception is thrown. You can use try-with-resources here, as we discussed (and I misremembered) earlier. 🙂

File: src/main/java/eventprocessing/DataReader.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+import com.amazonaws.services.s3.model.S3ObjectInputStream;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+
+public class DataReader {
+    public void saveToFile(S3ObjectInputStream amazonS3Stream) throws IOException {
+        FileOutputStream fos = new FileOutputStream(new File("locations.json"));
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/DataReader.java]: "locations.json" is hardcoded in a couple of different places. It's also easy to mix up with the S3 key. What about moving into the config?

File: src/main/java/eventprocessing/Main.java [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+            counter++;
+            if (counter > duration) {
+                sqsClient.destroyQueue();
+                break;
+            }
+        }
+    }
+
+    private static void createObjects() throws IOException {
+        int MAX_SIZE = 300;
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/Main.java]: In Java, you tend to only use `SCREAMING_SNAKE_CASE` for constants defined in the body of the class, rather than within a method.

That said, I'd suggest moving this to the config, but it might good to think about whether capping the list by size is the best thing to do - you don't always know what the message volume will be, so you might delete items too soon.

File: src/test/java/ResponseProcessorTest.java [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+    private Sensor mockSensor;
+    private SensorList mockSensorList;
+    private Message mockMessage;
+    private Response mockResponse;
+    private MessageLog mockMessageLog;
+    private ResponseProcessor responseProcessor;
+
+    @Before
+    public void setup() {
+        mockSensor = Mockito.mock(Sensor.class);
```
Comment (suzannehamilton) [inline src/test/java/ResponseProcessorTest.java]: This long setup in `@Before` makes it hard to follow the individual tests. For example, if you compare the tests on line 43 and line 49, it's not obvious _why_ `validId` returns `true`, and `invalidId` returns `false`.

Lots of shared setup also leads to fragile tests - you tweak some setup to add a new test, and it breaks an existing one. Or (worse) an existing one now passes for the wrong reason.

Try to put all the setup for a test within the test itself. Sometimes you can't avoid it, because the tests would be too slow if they did their own setup, but that normally only applies to integration tests that are slow because they set up database fixtures.

You can create helper methods like builders to simplify test setup, and you can design those so that the test setup is still obvious (i.e. without hiding data).

If your tests are still too long, it's usually a sign that the code under test has a complicated interface, and it should be simplified.

**Review by suzannehamilton — COMMENTED** [review]
I've finished reviewing this now. The approach looks good, and you've figured out an impressive amount of AWS wrangling without my help. 👍

File: src/main/java/eventprocessing/responseservices/ResponseService.java (hunk 8 above)
Comment (julesnuggy) [reply src/main/java/eventprocessing/responseservices/ResponseService.java]: Unfortunately not because the form we received it is not valid JSON. The field we want to extract - "message" - is not a valid JSON but is structured as one.
